presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+import com.facebook.presto.spi.block.VariableWidthBlock;
+import com.facebook.presto.spi.type.Type;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.array.Arrays.ensureCapacity;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DATA;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.LENGTH;
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.PRESENT;
+import static com.facebook.presto.orc.reader.SliceSelectiveReader.computeTruncatedLength;
+import static com.facebook.presto.orc.reader.SliceSelectiveReader.getMaxCodePointCount;
+import static com.facebook.presto.orc.stream.MissingInputStreamSource.missingStreamSource;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SliceDirectSelectiveStreamReader
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SliceDirectSelectiveStreamReader.class).instanceSize();
+    private static final int ONE_GIGABYTE = toIntExact(new DataSize(1, GIGABYTE).toBytes());
+
+    private final TupleDomainFilter filter;
+    private final boolean nonDeterministicFilter;
+    private final boolean nullsAllowed;
+
+    private final StreamDescriptor streamDescriptor;
+    private final Type outputType;
+
+    private int readOffset;
+
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    private BooleanInputStream presentStream;
+    private InputStreamSource<ByteArrayInputStream> dataStreamSource = missingStreamSource(ByteArrayInputStream.class);
+    private ByteArrayInputStream dataStream;
+    private InputStreamSource<LongInputStream> lengthStreamSource = missingStreamSource(LongInputStream.class);
+    private LongInputStream lengthStream;
+
+    private boolean rowGroupOpen;
+    private LocalMemoryContext systemMemoryContext;
+
+    private boolean[] nulls;
+
+    protected int[] outputPositions;
+    protected int outputPositionCount;
+
+    protected final boolean outputRequired;
+    protected boolean allNulls = true;
+    private boolean[] isNullVector;
+    private int[] lengthVector;
+    private int[] offsetVector;
+    private int totalLength;
+    private boolean valuesInUse;
+    private byte[] data;
+    private Slice slice;
+    private int lengthIndex;
+    private int maxLength;
```

Comment:
- maxLength and totalLength can be local variables in prepareForNextRead
- use `ensureCapacity` instead of `new` in prepareForNextRead to reuse arrays
- add some empty lines to make prepareForNextRead easier to read

```
    private void prepareForNextRead(int positionCount, int[] positions)
            throws IOException
    {
        lengthIndex = 0;
        outputPositionCount = 0;

        int totalLength = 0;
        int maxLength = 0;

        int totalPositions = positions[positionCount - 1] + 1;
        int nullCount = 0;
        if (presentStream != null) {
            isNullVector = ensureCapacity(isNullVector, totalPositions);
            nullCount = presentStream.getUnsetBits(totalPositions, isNullVector);
        }
        
        if (lengthStream != null) {
            lengthVector = ensureCapacity(lengthVector, totalPositions - nullCount);
            lengthStream.nextIntVector(totalPositions - nullCount, lengthVector, 0);

            offsetVector = ensureCapacity(offsetVector, totalPositions + 1);

            for (int i = 0; i < totalPositions - nullCount; i++) {
                totalLength += lengthVector[i];
                maxLength = Math.max(maxLength, lengthVector[i]);
            }

            if (totalLength > ONE_GIGABYTE) {
                throw new PrestoException(
                        GENERIC_INTERNAL_ERROR,
                        format("Values in column \"%s\" are too large to process for Presto. %s column values are larger than 1GB [%s]",
                                streamDescriptor.getFieldName(), positionCount,
                                streamDescriptor.getOrcDataSourceId()));
            }
        }

        if (outputRequired) {
            if (presentStream != null && nullsAllowed) {
                nulls = ensureCapacity(nulls, positionCount);
            }
            data = ensureCapacity(data, totalLength);
        }
        else {
            data = ensureCapacity(data, maxLength);
        }
        
        dataAsSlice = Slices.wrappedBuffer(data);
    }
```